### PR TITLE
[Merged by Bors] - feat: reimplement deepMap and deepMapKeys to fix vite build (VF-000)

### PR DESCRIPTION
**Pull request by z4o4z**
**Fixes or implements VF-000**

### Brief description. What is this change?

looks like one of the dependencies of `deepMap`, `deepMapKeys` or `es6-weak-map` is written in the common-js only way and Vite can't build it correctly. Moreover, we don't need all these polyfills that are used in the packages above since we support only modern browsers, and something really new can be polyfilled by bundler

### Checklist

- [ ] this is a breaking change and should publish a new major version
- [ ] appropriate tests have been written

### CODE REVIEW

File: packages/common/src/utils/array.ts [hunk 1]
```diff
@@ -65,25 +65,30 @@ export const separate = <T>(items: T[], predicate: (item: T, index: number) => b
...
-    (acc, item) => {
-      if (lhs.includes(item)) {
-        if (rhs.includes(item)) {
-          acc.union.push(item);
-        } else {
-          acc.lhsOnly.push(item);
-        }
+  const lMap = new Map(lhs.map((item) => [item, true]));
+  const rMap = new Map(rhs.map((item) => [item, true]));
+  const unionMap = new Map([...lMap.entries(), ...rMap.entries()]);
```
Comment (z4o4z) [inline packages/common/src/utils/array.ts:74]: this approach is a bit faster than nested .includes calls

File: packages/common/src/utils/object/common.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    return newObj;
+  }
+
+  if (keys.length === 2) {
+    delete newObj[keys[0]];
+    delete newObj[keys[1]];
+
+    return newObj;
+  }
```
Comment (z4o4z) [inline packages/common/src/utils/object/common.ts:37]: small optimizations, to do not call reduce just for 1 or 2 keys

File: packages/common/src/utils/object/common.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    return {} as Pick<T, K>;
+  }
+
+  if (keys.length === 1) {
+    return { [keys[0]]: obj[keys[0]] } as Pick<T, K>;
+  }
+
+  if (keys.length === 2) {
+    return { [keys[0]]: obj[keys[0]], [keys[1]]: obj[keys[1]] } as Pick<T, K>;
+  }
```
Comment (z4o4z) [inline packages/common/src/utils/object/common.ts:57]: same here

File: packages/common/src/utils/object/diff.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
+import { AnyRecord, DeepPartial } from '@common/types';
```
Comment (z4o4z) [inline packages/common/src/utils/object/diff.ts:1]: there're no changes, just moved into separate file

File: packages/common/src/utils/object/common.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
+import { AnyRecord, Struct } from '@common/types';
```
Comment (z4o4z) [inline packages/common/src/utils/object/common.ts:1]: there're mostly no changes, just moved into a separate file

File: packages/common/src/utils/object/keys.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
+import { AnyRecord } from '@common/types';
```
Comment (z4o4z) [inline packages/common/src/utils/object/keys.ts:1]: there're mostly no changes, just moved into a separate file

**Review by jonahsnider — APPROVED** [review]

File: packages/common/src/utils/array.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/common/src/utils/array.ts:74]: I'm using maps just for performance reason, here's the benchmarks: https://www.measurethat.net/Benchmarks/Show/17271/0/findunion---nested-arrayincludes-vs-maphas-vs-sethas

but yeah, going to replace `unionMap` to `unionSet` - works faster and makes more sense

File: packages/common/src/utils/array.ts [hunk 7]
```diff
@@ -65,25 +65,30 @@ export const separate = <T>(items: T[], predicate: (item: T, index: number) => b
 export const createMap = <T extends AnyRecord, K extends string | number = string>(array: T[], getKey: (value: T) => string) =>
   array.reduce<Record<K, T>>((acc, item) => Object.assign(acc, { [getKey(item)]: item }), {} as Record<K, T>);
 
+export const createPrimitiveMap = <T extends string | number | symbol>(array: T[]): Record<T, T> =>
+  array.reduce<Record<T, T>>((acc, item) => Object.assign(acc, { [item]: item }), {} as Record<T, T>);
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/common/src/utils/array.ts:69]: oh nice, didn't know about `PropertyKey`, thanks!
and yeah, `{  [P in T]: P; }` seems much better, going to add a type helper for this

File: packages/common/src/utils/array.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/common/src/utils/array.ts:74]: yeah, looks like we can replace it with `Set`, should be a bit faster.

File: packages/common/src/utils/array.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/common/src/utils/array.ts]: refactored `createMap` helper to work with primitives as well as objects

**Pull request closed without merging**